### PR TITLE
Fix halo again

### PR DESCRIPTION
**Pull request by ASKabalan**
This PR comes after this one https://github.com/DifferentiableUniverseInitiative/jaxDecomp/pull/19
Instead of allocating using `mlir.full_like_aval` I allocate using `cudecompMalloc` but I keep a fallback to XLA if needed using `JD_ALLOCATE_WITH_XLA=1`
@aboucaud @EiffL Good for review after merging https://github.com/DifferentiableUniverseInitiative/jaxDecomp/pull/19

### CODE REVIEW

**Comment by aboucaud** [comment]
Out of my league..

**Comment by ASKabalan** [comment]
Still much simpler than JAX distributed :smiling_face_with_tear:

**Comment by EiffL** [comment]
Hum... but I don't understand what was the issue here, as long as the workspace is created by XLA in each call the primitive and that the primitive uses the provided workspace in the input stream, what goes wrong?

**Comment by ASKabalan** [comment]
The workspace is not created at each call.
It is only created at the lowering phase as ir.constants

If I do an LPT it is fine.
If I do a nbody sim with a mesh size of 256 it is also fine (on a V100 32G)
512 and above I get a deadlock, with some further inspection it apeared to me that there is an illegal access of memory by some of the processes (which caused the deadlock)

I inspected the memory and found :

```
Calling cudecompUpdateHalosX : 0x14a75f004300 On axis 0
Calling cudecompUpdateHalosX : 0x14a75f004300 On axis 1
Calling cudecompUpdateHalosX : 0x14a75f004300 On axis 2

Calling cudecompUpdateHalosX : 0x151003004300 On axis 0
Calling cudecompUpdateHalosX : 0x151003004300 On axis 1
Calling cudecompUpdateHalosX : 0x151003004300 On axis 2
```

XLA changes the memory address which should not happen (again the lowering happends only once)
This means that the memory have been repurposed.

Using `cudecompmalloc` this does not happend.

You can still reproduce if you get a 512 or 1024 on no more than 4 GPUs with the flag `JD_ALLOCATE_WITH_XLA`

**Comment by ASKabalan** [comment]
To get the error to show, you need to remove all `block_until_ready() `statements

**Comment by EiffL** [comment]
Hum, but the thing that worries me is that the same could happen for the
other ops no?
And I don't see why the memory provided in the stream to the kernel would
not be ok.

**Comment by EiffL** [comment]
Are you sure it's not something different like not having the right size of
workspace? Maybe we have a different meaning for halo size

**Comment by ASKabalan** [comment]
I am going to try to reproduce and tell you.
But the halo_extent here is half of the halo size but what I see is there is enough space
The size of the input is mesh + half of the halo and the other half is the extent

this should be enough

For the FFT, the allocation is handled by us and not cufft because we use [cufftsetautoallocation](https://docs.nvidia.com/cuda/cufft/#cufftsetautoallocation) and the work is shared for the FFT and Transpose

The only difference I see is that the FFT is inplace (it uses some temporary parameters but nothing dangerous)
The transpositions are inplace aswell

The supplement memory required by a halo exchange is more?
Anyway it cannot be anything else, since when I use `cudecomp` the illegal access is not reproduced

File: jaxdecomp/_src/halo.py [hunk 1]
```diff
@@ -122,6 +123,11 @@ def lowering(ctx, x: Array, halo_extents: Tuple[int, int, int],
         config, is_double, halo_extents[::-1], halo_periods[::-1], 0)
     layout = tuple(range(n - 1, -1, -1))
 
+    # If XLA is not the selected allocator, then allocate a workspace with size 1
+    # TODO(Wassim) : Eventually, Only cuda should allocate .. this will be removed in the future
+    if os.environ.get("JD_ALLOCATE_WITH_XLA", "0") == "0":
```
Comment (EiffL) [inline jaxdecomp/_src/halo.py:128]: can we remove this entire mechanism?

File: src/halo.cu [hunk 2]
```diff
@@ -24,23 +24,34 @@ HRESULT HaloExchange<real_t>::get_halo_descriptor(cudecompHandle_t handle, size_
...
 
+  // TODO(Wassim) Handle complex numbers
   int64_t dtype_size;
   if (halo_desc.double_precision)
     CHECK_CUDECOMP_EXIT(cudecompGetDataTypeSize(CUDECOMP_DOUBLE, &dtype_size));
   else
     CHECK_CUDECOMP_EXIT(cudecompGetDataTypeSize(CUDECOMP_FLOAT, &dtype_size));
 
-  work_size = dtype_size * workspace_num_elements;
+  m_WorkSize = dtype_size * workspace_num_elements;
```
Comment (EiffL) [inline src/halo.cu:39]: let's remove this

**Comment by EiffL** [comment]
I think it's better to let XLA handle memory, because it can decide to free it when not necessary.

**Review by EiffL — CHANGES_REQUESTED** [review]
We should remove the cuda memory allocation thing for halos, better to be consistent in memory handling throughout

**Comment by ASKabalan** [comment]
I just tested.
It is ok for me.
Can I merge?

**Pull request merged**